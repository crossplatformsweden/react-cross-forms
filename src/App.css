.App {
  text-align: center;
  font-size: large;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  padding-top: 10vh;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

.App-form {
  max-width: 400pt;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.buttons {
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  flex-direction: row;
}

.countries{
  width: 150pt;
}

input, select, label {
  display: flex;
  flex-direction: row;
  margin-top: 5pt;
  align-items: center;
  justify-content: center;
  font-size: large;
}

button {
  display: flex;
  flex-direction: row;
  margin-top: 10pt;
  font-size: large;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
