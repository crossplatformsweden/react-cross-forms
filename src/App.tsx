import './App.css';

import { Form, Formik } from 'formik';
import React, { useState } from 'react';
import { useEffect } from 'react';
import Button from 'react-bootstrap/Button';
import * as Yup from 'yup';

import { FormField } from './components/FormField';
import { Country } from './Country';
import logo from './logo.svg';

interface IUserFields {
  phone: any;
  social: any;
  email: any;
  password: any;
  country: any;
}
const UserFieldKeys: IUserFields = {
  phone: 'phone',
  social: 'social',
  email: 'email',
  password: 'password',
  country: 'country',
};

const InitialValues: IUserFields = {
  email: '',
  password: '',
  social: '',
  phone: '',
  country: '',
};

const ValidationRules: IUserFields = {
  country: Yup.string()
    .required('Landskod krävs')
    .uppercase('Landskod i versaler')
    .length(2, 'Landskod med 2 tecken'),
  phone: Yup.string()
    .required('Telefon krävs')
    .matches(/^((((0{2}?)|(\+){1})46)|0)[\d]{9}/, 'Ogiltigt telefonnummer'),
  social: Yup.string()
    .required('Fältet krävs')
    .matches(/^(19|20)?[0-9]{6}[- ]?[0-9]{4}$/, 'Ogiltigt personnummer'),
  email: Yup.string()
    .email('Ogiltig e-postadress')
    .required('E-postadress krävs'),
  password: Yup.string()
    .min(5, 'Lösenord måste vara minst 5 tecken')
    .required('Lösenord krävs'),
};

export const App: React.FunctionComponent = () => {
  const [countries, setCountries] = useState<Country[] | undefined>(undefined);

  useEffect(() => {
    const getCountries = async () => {
      try {
        const local = localStorage.getItem(UserFieldKeys.country);
        if (local?.length) {
          setCountries(JSON.parse(local));
        }
        const response = await fetch('https://restcountries.eu/rest/v2/all');
        if (!response || !response.ok) {
          return;
        }

        const resultText = await response.text();
        const newCountries: Country[] = JSON.parse(resultText) as Country[];
        if (newCountries) {
          localStorage.setItem(
            UserFieldKeys.country,
            JSON.stringify(newCountries)
          );
          setCountries(newCountries);
        }
      } catch (error) {
        alert('Kunde inte hämta länder');
      }
    };
    getCountries();
  }, []);
  return (
    <div className="App">
      <header className="App-header">
        <img title="logo" src={logo} className="App-logo" alt="logo" />

        <div className="App-form">
          <Formik
            initialValues={InitialValues}
            validateOnBlur
            validationSchema={Yup.object(ValidationRules)}
            onSubmit={(
              values,
              { setSubmitting, setErrors, setStatus, resetForm }
            ) => {
              try {
                setTimeout(() => {
                  console.log('Success');
                  if (
                    process.env.NODE_ENV ||
                    process.env.NODE_ENV === 'development'
                  ) {
                    console.log(JSON.stringify(values, null, 2));
                    alert('Tack!');
                  }
                  setSubmitting(false);
                }, 400);
                resetForm({});
                setStatus({ success: true });
              } catch (error) {
                setStatus({ success: false });
                setSubmitting(false);
                // setErrors({ submit: error.message });
              }
            }}
          >
            {({ isSubmitting }) => (
              <Form>
                <FormField
                  name={UserFieldKeys.social}
                  placeholder="ÅÅMMDD-XXXX"
                  label="Personnummer"
                />
                <FormField
                  type="email"
                  autoComplete="username"
                  name={UserFieldKeys.email}
                  placeholder="E-post"
                />
                <FormField
                  type="phone"
                  name={UserFieldKeys.phone}
                  placeholder="Telefonnummer"
                />
                <FormField
                  type="password"
                  autoComplete="new-password"
                  name={UserFieldKeys.password}
                  placeholder="Lösenord"
                />
                <div>
                  {countries ? (
                    <FormField
                      name={UserFieldKeys.country}
                      as="select"
                      className="countries"
                    >
                      {countries.map((c) => (
                        <option
                          key={c.alpha2Code || '1'}
                          value={c.alpha2Code || ''}
                        >
                          {c.name} ({c.alpha2Code}) +{c.callingCodes}{' '}
                          {c.topLevelDomain}
                        </option>
                      ))}
                    </FormField>
                  ) : null}
                </div>
                <div className="buttons">
                  <Button
                    type="submit"
                    disabled={isSubmitting}
                    variant="success"
                  >
                    Submit
                  </Button>
                  <Button
                    type="reset"
                    variant="danger"
                    onReset={() => localStorage.clear()}
                  >
                    Rensa
                  </Button>
                </div>
              </Form>
            )}
          </Formik>
        </div>
      </header>
    </div>
  );
};
