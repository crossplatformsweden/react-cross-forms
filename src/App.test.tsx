import React from 'react';
import renderer from 'react-test-renderer';
import { App } from './App';

describe('components', () => {
  describe('<App />', () => {
    test('renders correctly', () => {
      const instance = renderer.create(<App onSubmit={() => {}} />).toJSON();
      expect(instance).toMatchSnapshot();
    });

    test('logo has correct class', () => {
      const instance = renderer.create(<App onSubmit={() => {}} />);
      const child = instance?.root.findByProps({ title: 'logo' });
      expect(child.props.className).toBe('App-logo');
    });
  });
});
