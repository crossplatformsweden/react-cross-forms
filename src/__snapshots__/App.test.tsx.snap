// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <App /> renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="test-file-stub"
      title="logo"
    />
    <div
      className="App-form"
    >
      <form
        action="#"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div>
          <input
            name="social"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="ÅÅMMDD-XXXX"
            value=""
          />
        </div>
        <div>
          <input
            autoComplete="username"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="E-post"
            type="email"
            value=""
          />
        </div>
        <div>
          <input
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Telefonnummer"
            type="phone"
            value=""
          />
        </div>
        <div>
          <input
            autoComplete="new-password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Lösenord"
            type="password"
            value=""
          />
        </div>
        <div />
        <div
          className="buttons"
        >
          <button
            className="btn btn-success"
            disabled={false}
            type="submit"
          >
            Submit
          </button>
          <button
            className="btn btn-danger"
            disabled={false}
            onReset={[Function]}
            type="reset"
          >
            Rensa
          </button>
        </div>
      </form>
    </div>
  </header>
</div>
`;
